### DISCUSSION TITLE
详解分布式协调服务 ZooKeeper · /zookeeper-chubby

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#530] [Announcements]
https://draveness.me/zookeeper-chubby

**Comment by XhinLiang** [comment]
常见的通知机制往往都有两种，一种是客户端使用『拉』的方式从服务端获取最新的状态，这种方式获取的状态很有可能都是过期的，需要客户端不断地通过轮训的方式获取服务端最新的状态，另一种方式就是在客户端订阅对应节点后由服务端向所有订阅者推送该节点的变化，相比于客户端主动获取数据的方式，服务端主动推送更能够保证客户端数据的实时性。

轮训 --> 轮询

**Comment by draveness** [comment]
> @XhinLiang
> 常见的通知机制往往都有两种，一种是客户端使用『拉』的方式从服务端获取最新的状态，这种方式获取的状态很有可能都是过期的，需要客户端不断地通过轮训的方式获取服务端最新的状态，另一种方式就是在客户端订阅对应节点后由服务端向所有订阅者推送该节点的变化，相比于客户端主动获取数据的方式，服务端主动推送更能够保证客户端数据的实时性。
> 
> 轮训 --> 轮询

fixed

**Comment by inverai** [comment]
相比于名不见经传的 Chubby，作者相信 Zookeeper 更被广大开发者所熟知，作为非常出名的分布式协调服务，Zookeeper 有非常多的应用，包括发布订阅、命名服务、分数是协调和分布式锁，

分数是协调和分布式锁

**Comment by draveness** [comment]
@inverai 修复了

**Comment by taojintianxia** [comment]
图是用什么工具画的 ? 看着不错.

**Comment by 7u** [comment]
```java
Integer number = Integer.parseInt(lockNumber.replace("/resource/lock-", "")) + 1;
```
如果是监听前一个节点的删除事件，这里是不是要 - 1 ?

**Comment by Michael-CX** [comment]
请教一个问题：
Zookeeper写数据时，写到超过一半副本就算成功。
如果一个查询数据的请求，落到旧数据的机器上，是不是读到的就是旧数据呢？

**Comment by wadekun** [comment]
> @7u
> ```java
> Integer number = Integer.parseInt(lockNumber.replace("/resource/lock-", "")) + 1;
> ```
> 如果是监听前一个节点的删除事件，这里是不是要 - 1 ?

应该是减1。。

**Comment by draveness** [comment]
@7u @wadekun 确实写错了...我改下

**Comment by yimkong** [comment]
分布式锁，如果选择监听前一个节点的时候，前一个节点成功执行并且被删除了，监听一个不存在的节点，会出现监听不到的情况吧，这也是并发的一个问题。

**Comment by draveness** [comment]
> 分布式锁，如果选择监听前一个节点的时候，前一个节点成功执行并且被删除了，监听一个不存在的节点，会出现监听不到的情况吧，这也是并发的一个问题。

前一节点执行成功并删除了，后一个节点会收到删除的通知

**Comment by lintiyan** [comment]
提两个看了文章后有疑惑的问题：

1. zookeeper的主节点在半数从节点返回确认后，确认提交更改。如果A客户端更改了某个path值，主节点收到半数从节点返回的确认时，还有小半数的从节点没有同步。这时，B客户端请求获取这个path值，且这个请求落在了那小半数未同步的节点了，这时数据不是不一致了么，B拿到的还是旧的数据。 这个是怎么处理的？

2. 通知机制里，文章里只是提了获取节点值时，可以注册一个带回调方法process的watcher，当节点值修改时会触发这个watcher的process方法，可是这个触发是怎么实现的？更改是如何触发process的呢？

大佬解惑呀！

**Comment by lintiyan** [comment]
找到答案了，自己来答一波。

zookeeper的数据同步过程很类似分布式事务的两阶段提交。leader收到数据更新操作时，将更新信息发送给follower，follower接受到信息并且以事务的形式将更新信息落地后，返回给leader一个ack，当leader接收到超过半数follower的ack时，leader就向follower发送事务提交消息，follower接受事务提交信息后执行事务提交。  

所以是我理解错了，不是半数follower都提交后，leader进行提交；而是半数follower返回ack后，leader发送提交，表示同步完成。
如果又一个follower的本地事务执行失败，leader会让所有follower执行回滚，也就是这个更改在zookeeper中更新失败。所以follower中的数据总是一致的

**Comment by Jigsawk** [comment]
发布订阅的那个demo，感觉有一个细节需要注意一下，watcher是一次性的，所以需要在回调方法内再次注册一个新的watcher，才能监听到该节点后续的变动

**Comment by xiaoyunyu** [comment]
> @yimkong
> 分布式锁，如果选择监听前一个节点的时候，前一个节点成功执行并且被删除了，监听一个不存在的节点，会出现监听不到的情况吧，这也是并发的一个问题。


可以读取 zk.getDatade 的响应, 如果为空表示前一个节点已经被删除了, 本节点已经获取锁

**Comment by mmoly790** [comment]
think

**Reply by Uyouii** [reply]
学习
